toothpick-runtime/src/main/java/toothpick/Toothpick.java

```diff
@@ -39,12 +39,31 @@
   // its transformation
   private static final ConcurrentHashMap<Object, Scope> MAP_KEY_TO_SCOPE =
       new ConcurrentHashMap<>();
+  private static final ConcurrentHashMap<Object, Scope> ROOT_SCOPES = new ConcurrentHashMap<>();
   static Injector injector = new InjectorImpl();
 
   protected Toothpick() {
     throw new RuntimeException("Constructor can't be invoked even via reflection.");
   }
 
+  /**
+   * @return The root scope as follows: <br>
+   *     <ul>
+   *       <li>If there's an existing scope tree the root scope of that tree
+   *       <li>If there's no tree it will create a default root scope and return it
+   *       <li>If there are multiple scope trees (e.g. a forest) this method will throw an exception
+   *     </ul>
+   */
+  public static Scope openRootScope() {
```

Comment:
this is not thread safe, we will have to synchnized this and also the place where we open the scopes (line 157):

```java
  private static Scope openScope(Object name, boolean shouldCheckMultipleRootScopes) {
    if (name == null) {
      throw new IllegalArgumentException("null scope names are not allowed.");
    }

    Scope scope = MAP_KEY_TO_SCOPE.get(name);
    if (scope != null) {
      return scope;
    }
    scope = new ScopeImpl(name);
    Scope previous = MAP_KEY_TO_SCOPE.putIfAbsent(name, scope); //  <----- HERE
    if (previous != null) {
      // if there was already a scope by this name, we return it
      scope = previous;
    } else if (shouldCheckMultipleRootScopes) {
      ConfigurationHolder.configuration.checkMultipleRootScopes(scope);
    }
    return scope;
  }
```

Reply:
Ideally try to synchronized if we have to open the scopes, not always. (similar to double-check for singletons)

Reply:
@dlemures can we do the synchronisation in a separate PR? that is a broader effort, not necessarily related to this change.

Reply:
The issue is that Toothpick is right now thread safe:
```java
 MAP_KEY_TO_SCOPE.putIfAbsent(name, scope); <--- This line alone is thread safe
```
But this change is breaking the safety. I think it would be nice not to merge it to master with concurrency working.

Reply:
so we should synchronise openRootScope and openScope? I'm not sure I understand what method of synchronisation you would like here...

Reply:
We want to sync on the new map key set because size and put / get should be atomic. 
We should just merge when sync is done right.

Reply:
@dlemures can you tell me exactly the race condition that will cause an issue. I think we're ok with how openScope() already works.

Reply:
Hey @code-twister sorry for the late reply, this week I will be off so I might not be able to reply.

The point is that whenever someone calls `openScope/s`(explicit root scope) from one thread together with `openRootScope` from another thread around the same time, there are 2 possibilities:

1. `openScope/s` is executed first and then `openRootScope` will open that same Scope.
2. `openRootScope` is executed first and then `openScope/s` will happen next and open a new scope (or crash if we have the configuration for it to happen).

That is a valid behavior (race condition) and TP is working as it should, it is the user responsibility to be aware of how `openRootScope` works (also we should make it very clear in the doc). So there is no issue there.

However, the current implementation there are situation where this method will not behave as expected:

**Example 1**
1. Initial situation: 0 root scopes
2. Thread 1:  finishes checking both `if statement` with false inside `openRootScope`.
3. Thread 2: opens new root scope with a custom name
4. Thread 1: opens a second root scope with name `Toothpick.class` and does not fail.

**Example 2**
1. Initial situation: 1 root scope
2. Thread 1: inside `} else if (ROOT_SCOPES.size() == 1) {` block
3. Thread 2: executes line `ROOT_SCOPES.remove(name);`
4. Thread 1: throws `NoSuchElementException` as the process was closed in the mean time

There might be more scenarios. My recommendation is to think about all possible combinations. It is possible that we might have something like:

**Valid race condition**
1. Initial situation: 1 root scope
2. Thread 1: inside `} else if (ROOT_SCOPES.size() == 1) {` block
3. Thread 2: opens a second root scope (check off)
4. Thread 1: will not fail and just return the first root scope from the map `ROOT_SCOPES`

But I consider that it is fine, as it is as if the user just opened the root scope after `openRootScope` is opened. And that doesnt happen the contract of the method/TP. However, the `example 1` above does as it is `openRootScope` the one opening the second scope.

Sorry for the long comment, i hope it is clear.

Reply:
It's clear... thanks @dlemures!
I will implement a solution soon.

Reply:
Actually, I think that **example 1** could be alright. It would be kind of equivalent to:

> openRootScope is executed first and then openScope/s will happen next and open a new scope (or crash if we have the configuration for it to happen).